serialize_buffer allocator support needs adjustments

The `serialize_buffer` allocator support never calls the destructor of the wrapped buffers, it only invokes `Allocator::deallocate`. This also causes the wrong deallocation function to be called for array types (`operator delete(T*)` instead of `operator delete[](T*)`).

Example error message generated by gcc (release):
```txt
/work/jenkins/workspace/HPX/hpx/examples/quickstart/zerocopy_rdma.cpp
In file included from /opt/apps/gcc/12.3.0/include/c++/12.3.0/x86_64-redhat-linux/bits/c++allocator.h:33,
                 from /opt/apps/gcc/12.3.0/include/c++/12.3.0/bits/allocator.h:46,
                 from /opt/apps/gcc/12.3.0/include/c++/12.3.0/unordered_map:40,
                 from /opt/apps/gcc/12.3.0/include/c++/12.3.0/functional:61,
                 from /work/jenkins/workspace/HPX/hpx/libs/core/tag_invoke/include/hpx/functional/detail/invoke.hpp:11,
                 from /work/jenkins/workspace/HPX/hpx/libs/core/functional/include/hpx/functional/invoke.hpp:10,
                 from /work/jenkins/workspace/HPX/hpx/libs/core/algorithms/include/hpx/parallel/algorithms/adjacent_difference.hpp:223,
                 from /work/jenkins/workspace/HPX/hpx/libs/core/algorithms/include/hpx/parallel/algorithm.hpp:17,
                 from /work/jenkins/workspace/HPX/hpx/libs/full/include/include/hpx/algorithm.hpp:10,
                 from /work/jenkins/workspace/HPX/hpx/libs/full/include/include/hpx/hpx.hpp:9,
                 from /work/jenkins/workspace/HPX/hpx/examples/quickstart/zerocopy_rdma.cpp:9:
In member function â€˜void std::__new_allocator<_Tp>::deallocate(_Tp*, size_type) [with _Tp = double]â€™,
    inlined from â€˜constexpr void std::allocator< <template-parameter-1-1> >::deallocate(_Tp*, std::size_t) [with _Tp = double]â€™ at /opt/apps/gcc/12.3.0/include/c++/12.3.0/bits/allocator.h:200:35,
    inlined from â€˜static void hpx::serialization::serialize_buffer<T, Allocator>::deleter(T*, Deallocator, std::size_t) [with Deallocator = std::allocator<double>; T = double; Allocator = std::allocator<double>]â€™ at /work/jenkins/workspace/HPX/hpx/libs/core/serialization/include/hpx/serialization/serialize_buffer.hpp:47:31,
    inlined from â€˜hpx::serialization::serialize_buffer<double>::serialize_buffer(double*, std::size_t, init_mode, const allocator_type&)::<lambda(double*)>â€™ at /work/jenkins/workspace/HPX/hpx/libs/core/serialization/include/hpx/serialization/serialize_buffer.hpp:108:66,
    inlined from â€˜std::__shared_count<_Lp>::__shared_count(_Ptr, _Deleter, _Alloc) [with _Ptr = double*; _Deleter = hpx::serialization::serialize_buffer<double>::serialize_buffer(double*, std::size_t, init_mode, const allocator_type&)::<lambda(double*)>; _Alloc = std::allocator<void>; <template-parameter-2-4> = void; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]â€™ at /opt/apps/gcc/12.3.0/include/c++/12.3.0/bits/shared_ptr_base.h:958:11,
    inlined from â€˜std::__shared_count<_Lp>::__shared_count(_Ptr, _Deleter) [with _Ptr = double*; _Deleter = hpx::serialization::serialize_buffer<double>::serialize_buffer(double*, std::size_t, init_mode, const allocator_type&)::<lambda(double*)>; <template-parameter-2-3> = void; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]â€™ at /opt/apps/gcc/12.3.0/include/c++/12.3.0/bits/shared_ptr_base.h:939:57,
    inlined from â€˜std::__shared_ptr<_Tp, _Lp>::__shared_ptr(_Yp*, _Deleter) [with _Yp = double; _Deleter = hpx::serialization::serialize_buffer<double>::serialize_buffer(double*, std::size_t, init_mode, const allocator_type&)::<lambda(double*)>; <template-parameter-2-3> = void; _Tp = double []; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]â€™ at /opt/apps/gcc/12.3.0/include/c++/12.3.0/bits/shared_ptr_base.h:1478:17,
    inlined from â€˜std::shared_ptr<_Tp>::shared_ptr(_Yp*, _Deleter) [with _Yp = double; _Deleter = hpx::serialization::serialize_buffer<double>::serialize_buffer(double*, std::size_t, init_mode, const allocator_type&)::<lambda(double*)>; <template-parameter-2-3> = void; _Tp = double []]â€™ at /opt/apps/gcc/12.3.0/include/c++/12.3.0/bits/shared_ptr.h:232:48,
    inlined from â€˜hpx::serialization::serialize_buffer<T, Allocator>::serialize_buffer(T*, std::size_t, init_mode, const allocator_type&) [with T = double; Allocator = std::allocator<double>]â€™ at /work/jenkins/workspace/HPX/hpx/libs/core/serialization/include/hpx/serialization/serialize_buffer.hpp:106:25,
    inlined from â€˜int main()â€™ at /work/jenkins/workspace/HPX/hpx/examples/quickstart/zerocopy_rdma.cpp:235:57:
/opt/apps/gcc/12.3.0/include/c++/12.3.0/bits/new_allocator.h:158:33: error: â€˜void operator delete(void*, std::size_t)â€™ called on pointer returned from a mismatched allocation function [-Werror=mismatched-new-delete]
  158 |         _GLIBCXX_OPERATOR_DELETE(_GLIBCXX_SIZED_DEALLOC(__p, __n));
      |                                 ^
/work/jenkins/workspace/HPX/hpx/examples/quickstart/zerocopy_rdma.cpp: In function â€˜int main()â€™:
/work/jenkins/workspace/HPX/hpx/examples/quickstart/zerocopy_rdma.cpp:234:64: note: returned from â€˜void* operator new [](std::size_t)â€™
  234 |         general_buffer_type buffer(new double[ZEROCOPY_DATASIZE],
      |                                                                ^
```

--- a/examples/1d_stencil/1d_stencil_6.cpp
+++ b/examples/1d_stencil/1d_stencil_6.cpp
@@ -65,7 +65,7 @@ public:
 
     // Create a new (uninitialized) partition of the given size.
     explicit partition_data(std::size_t size)
-      : data_(std::allocator<double>().allocate(size), size, buffer_type::take)
+      : data_(new double[size], size, buffer_type::take)
       , size_(size)
       , min_index_(0)
     {
@@ -73,7 +73,7 @@ public:
 
     // Create a new (initialized) partition of the given size.
     partition_data(std::size_t size, double initial_value)
-      : data_(std::allocator<double>().allocate(size), size, buffer_type::take)
+      : data_(new double[size], size, buffer_type::take)
       , size_(size)
       , min_index_(0)
     {

--- a/examples/1d_stencil/1d_stencil_7.cpp
+++ b/examples/1d_stencil/1d_stencil_7.cpp
@@ -66,7 +66,7 @@ public:
 
     // Create a new (uninitialized) partition of the given size.
     explicit partition_data(std::size_t size)
-      : data_(std::allocator<double>().allocate(size), size, buffer_type::take)
+      : data_(new double[size], size, buffer_type::take)
       , size_(size)
       , min_index_(0)
     {
@@ -74,7 +74,7 @@ public:
 
     // Create a new (initialized) partition of the given size.
     partition_data(std::size_t size, double initial_value)
-      : data_(std::allocator<double>().allocate(size), size, buffer_type::take)
+      : data_(new double[size], size, buffer_type::take)
       , size_(size)
       , min_index_(0)
     {

--- a/examples/1d_stencil/1d_stencil_8.cpp
+++ b/examples/1d_stencil/1d_stencil_8.cpp
@@ -105,12 +105,12 @@ private:
 
     struct hold_reference
     {
-        hold_reference(buffer_type const& data)
+        explicit hold_reference(buffer_type const& data)
           : data_(data)
         {
         }
 
-        void operator()(double*) {}    // no deletion necessary
+        void operator()(double const*) const {}    // no deletion necessary
 
         buffer_type data_;
     };

--- a/examples/quickstart/zerocopy_rdma.cpp
+++ b/examples/quickstart/zerocopy_rdma.cpp
@@ -49,11 +49,11 @@ public:
     {
     }
 
-    pointer address(reference value) const
+    static pointer address(reference value)
     {
         return &value;
     }
-    const_pointer address(const_reference value) const
+    static const_pointer address(const_reference value)
     {
         return &value;
     }
@@ -118,7 +118,7 @@ private:
     }
 
 public:
-    zerocopy_server(std::size_t size = 0)
+    explicit zerocopy_server(std::size_t size = 0)
       : data_(size, 3.1415)
     {
     }
@@ -127,7 +127,7 @@ public:
     // Retrieve an array of doubles to the given address
     transfer_buffer_type get_here(std::size_t size, std::size_t remote_buffer)
     {
-        pointer_allocator<double> allocator(
+        pointer_allocator<double> const allocator(
             reinterpret_cast<double*>(remote_buffer), size);
 
         // lock the mutex, will be unlocked by the transfer buffer's deleter
@@ -225,9 +225,7 @@ public:
 ///////////////////////////////////////////////////////////////////////////////
 int main()
 {
-    std::vector<hpx::id_type> localities = hpx::find_all_localities();
-
-    for (hpx::id_type const& id : localities)
+    for (hpx::id_type const& id : hpx::find_all_localities())
     {
         zerocopy zc = hpx::new_<zerocopy_server>(id, ZEROCOPY_DATASIZE);
 
@@ -238,7 +236,10 @@ int main()
             hpx::chrono::high_resolution_timer t;
 
             for (int i = 0; i != 100; ++i)
-                zc.get(hpx::launch::sync, ZEROCOPY_DATASIZE);
+            {
+                [[maybe_unused]] auto r =
+                    zc.get(hpx::launch::sync, ZEROCOPY_DATASIZE);
+            }
 
             double d = t.elapsed();
             std::cout << "Elapsed time 'get' (locality "
@@ -252,7 +253,7 @@ int main()
             for (int i = 0; i != 100; ++i)
                 zc.get_here(hpx::launch::sync, buffer);
 
-            double d = t.elapsed();
+            double const d = t.elapsed();
             std::cout << "Elapsed time 'get_here' (locality "
                       << hpx::naming::get_locality_id_from_id(id) << "): " << d
                       << "[s]\n";

--- a/libs/core/serialization/include/hpx/serialization/serialize_buffer.hpp
+++ b/libs/core/serialization/include/hpx/serialization/serialize_buffer.hpp
@@ -20,11 +20,55 @@
 #include <boost/shared_array.hpp>
 #endif
 
-#include <algorithm>
 #include <cstddef>
 #include <memory>
 
 namespace hpx::serialization {
+
+    namespace detail {
+
+        template <typename Allocator>
+        struct array_allocator
+        {
+            auto* operator()(Allocator alloc, std::size_t size) const
+            {
+                auto* p = alloc.allocate(size);
+                std::uninitialized_default_construct(p, p + size);
+                return p;
+            }
+        };
+
+        template <typename T>
+        struct array_allocator<std::allocator<T>>
+        {
+            T* operator()(std::allocator<T>, std::size_t size) const
+            {
+                return new T[size];
+            }
+        };
+
+        template <typename Deallocator>
+        struct array_deleter
+        {
+            template <typename T>
+            void operator()(
+                T* p, Deallocator dealloc, std::size_t size) const noexcept
+            {
+                std::destroy(p, p + size);
+                dealloc.deallocate(p, size);
+            }
+        };
+
+        template <typename T>
+        struct array_deleter<std::allocator<T>>
+        {
+            void operator()(
+                T const* p, std::allocator<T>, std::size_t) const noexcept
+            {
+                delete[] p;
+            }
+        };
+    }    // namespace detail
 
     ///////////////////////////////////////////////////////////////////////////
     template <typename T, typename Allocator>
@@ -41,20 +85,28 @@ namespace hpx::serialization {
         static constexpr void no_deleter(T*) noexcept {}
 
         template <typename Deallocator>
-        static void deleter(
-            T* p, Deallocator dealloc, std::size_t size) noexcept
+        struct deleter
         {
-            dealloc.deallocate(p, size);
-        }
+            void operator()(
+                T* p, Deallocator dealloc, std::size_t size) const noexcept
+            {
+                std::destroy_at(p);
+                dealloc.deallocate(p, size);
+            }
+        };
 
     public:
         enum init_mode
         {
-            copy = 0,         // constructor copies data
-            reference = 1,    // constructor does not copy data and does not
-                              // manage the lifetime of it
-            take = 2          // constructor does not copy data but does take
-                              // ownership and manages the lifetime of it
+            copy = 0,          // constructor copies data
+            reference = 1,     // constructor does not copy data and does not
+                               // manage the lifetime of it
+            take = 2,          // constructor does not copy data but takes
+                               // ownership of array pointer and manages the
+                               // lifetime of it
+            take_single = 3    // constructor does not copy data but takes
+                               // ownership of pointer to non-array and manages
+                               // the lifetime of it
         };
 
         using value_type = T;
@@ -72,9 +124,10 @@ namespace hpx::serialization {
           , size_(size)
           , alloc_(alloc)
         {
-            data_.reset(alloc_.allocate(size),
-                [alloc = this->alloc_, size = this->size_](T* p) noexcept {
-                    serialize_buffer::deleter<allocator_type>(p, alloc, size);
+            auto* p = detail::array_allocator<allocator_type>()(alloc_, size);
+            data_.reset(
+                p, [alloc = this->alloc_, size = this->size_](T* p) noexcept {
+                    serialize_buffer::deleter<allocator_type>()(p, alloc, size);
                 });
         }
 
@@ -88,25 +141,33 @@ namespace hpx::serialization {
         {
             if (mode == copy)
             {
-                data_.reset(alloc_.allocate(size),
+                auto* p =
+                    detail::array_allocator<allocator_type>()(alloc_, size);
+                data_.reset(p,
                     [alloc = this->alloc_, size = this->size_](T* p) noexcept {
-                        serialize_buffer::deleter<allocator_type>(
-                            p, alloc, size);
+                        detail::array_deleter<allocator_type>()(p, alloc, size);
                     });
-                if (size != 0)
-                    std::copy(data, data + size, data_.get());
+                std::uninitialized_copy(data, data + size, p);
             }
             else if (mode == reference)
             {
                 data_ = buffer_type(data, &serialize_buffer::no_deleter);
             }
-            else
+            else if (mode == take_single)
             {
                 // take ownership
                 data_ = buffer_type(data,
                     [alloc = this->alloc_, size = this->size_](T* p) noexcept {
-                        serialize_buffer::deleter<allocator_type>(
+                        serialize_buffer::deleter<allocator_type>()(
                             p, alloc, size);
+                    });
+            }
+            else if (mode == take)
+            {
+                // take ownership
+                data_ = buffer_type(data,
+                    [alloc = this->alloc_, size = this->size_](T* p) noexcept {
+                        detail::array_deleter<allocator_type>()(p, alloc, size);
                     });
             }
         }
@@ -120,49 +181,60 @@ namespace hpx::serialization {
         {
             // if 2 allocators are specified we assume mode 'take'
             data_ = buffer_type(data, [this, dealloc](T* p) noexcept {
-                serialize_buffer::deleter<Deallocator>(p, dealloc, size_);
+                detail::array_deleter<Deallocator>()(p, dealloc, size_);
             });
         }
 
         template <typename Deleter>
         serialize_buffer(T* data, std::size_t size, init_mode mode,
-            Deleter const& deleter,
-            allocator_type const& alloc = allocator_type())
+            Deleter&& deleter, allocator_type const& alloc = allocator_type())
           : data_()
           , size_(size)
           , alloc_(alloc)
         {
             if (mode == copy)
             {
-                data_.reset(alloc_.allocate(size), deleter);
-                if (size != 0)
-                    std::copy(data, data + size, data_.get());
+                auto* p =
+                    detail::array_allocator<allocator_type>()(alloc_, size);
+                data_ = buffer_type(p,
+                    [alloc = this->alloc_, size = this->size_](T* p) noexcept {
+                        detail::array_deleter<allocator_type>()(p, alloc, size);
+                    });
+                std::uninitialized_copy(data, data + size, p);
             }
             else
             {
                 // reference or take ownership, behavior is defined by deleter
-                data_ = buffer_type(data, deleter);
+                data_ = buffer_type(data,
+                    [deleter = HPX_FORWARD(Deleter, deleter)](
+                        T* p) noexcept { deleter(p); });
             }
         }
 
         template <typename Deleter>
         serialize_buffer(T const* data, std::size_t size,
             init_mode mode,    //-V659
-            Deleter const& deleter,
-            allocator_type const& alloc = allocator_type())
+            Deleter&& deleter, allocator_type const& alloc = allocator_type())
           : data_()
           , size_(size)
           , alloc_(alloc)
         {
             if (mode == copy)
             {
-                data_.reset(alloc_.allocate(size), deleter);
-                if (size != 0)
-                    std::copy(data, data + size, data_.get());
+                auto* p =
+                    detail::array_allocator<allocator_type>()(alloc_, size);
+                data_ = buffer_type(p,
+                    [alloc = this->alloc_, size = this->size_](T* p) noexcept {
+                        detail::array_deleter<allocator_type>()(p, alloc, size);
+                    });
+                std::uninitialized_copy(data, data + size, p);
             }
             else if (mode == reference)
             {
-                data_ = buffer_type(const_cast<T*>(data), deleter);
+                // reference behavior is defined by deleter
+                data_ = buffer_type(const_cast<T*>(data),
+                    [deleter = HPX_FORWARD(Deleter, deleter)](
+                        T* p) noexcept { deleter(p); });
             }
             else
             {
@@ -193,26 +265,28 @@ namespace hpx::serialization {
           , alloc_(alloc)
         {
             // create from const data implies 'copy' mode
-            data_.reset(alloc_.allocate(size),
-                [alloc = this->alloc_, size = this->size_](T* p) noexcept {
-                    serialize_buffer::deleter<allocator_type>(p, alloc, size);
+            auto* p = detail::array_allocator<allocator_type>()(alloc_, size);
+            data_ = buffer_type(
+                p, [alloc = this->alloc_, size = this->size_](T* p) noexcept {
+                    detail::array_deleter<allocator_type>()(p, alloc, size);
                 });
-            if (size != 0)
-                std::copy(data, data + size, data_.get());
+            std::uninitialized_copy(data, data + size, p);
         }
 
         template <typename Deleter>
-        serialize_buffer(T const* data, std::size_t size,
-            Deleter const& deleter,
+        serialize_buffer(T const* data, std::size_t size, Deleter&&,
             allocator_type const& alloc = allocator_type())
           : data_()
           , size_(size)
           , alloc_(alloc)
         {
             // create from const data implies 'copy' mode
-            data_.reset(alloc_.allocate(size), deleter);
-            if (size != 0)
-                std::copy(data, data + size, data_.get());
+            auto* p = detail::array_allocator<allocator_type>()(alloc_, size);
+            data_ = buffer_type(
+                p, [alloc = this->alloc_, size = this->size_](T* p) noexcept {
+                    detail::array_deleter<allocator_type>()(p, alloc, size);
+                });
+            std::uninitialized_copy(data, data + size, p);
         }
 
         serialize_buffer(T const* data, std::size_t size, init_mode mode,
@@ -223,13 +297,13 @@ namespace hpx::serialization {
         {
             if (mode == copy)
             {
-                data_.reset(alloc_.allocate(size),
+                auto* p =
+                    detail::array_allocator<allocator_type>()(alloc_, size);
+                data_ = buffer_type(p,
                     [alloc = this->alloc_, size = this->size_](T* p) noexcept {
-                        serialize_buffer::deleter<allocator_type>(
-                            p, alloc, size);
+                        detail::array_deleter<allocator_type>()(p, alloc, size);
                     });
-                if (size != 0)
-                    std::copy(data, data + size, data_.get());
+                std::uninitialized_copy(data, data + size, p);
             }
             else if (mode == reference)
             {
@@ -316,9 +390,10 @@ namespace hpx::serialization {
         {
             ar >> size_ >> alloc_;    // -V128
 
-            data_.reset(alloc_.allocate(size_),
-                [alloc = this->alloc_, size = this->size_](T* p) {
-                    serialize_buffer::deleter<allocator_type>(p, alloc, size);
+            data_ = buffer_type(
+                detail::array_allocator<allocator_type>()(alloc_, size_),
+                [alloc = this->alloc_, size = this->size_](T* p) noexcept {
+                    detail::array_deleter<allocator_type>()(p, alloc, size);
                 });
 
             if (size_ != 0)

--- a/libs/full/checkpoint/examples/1d_stencil_4_checkpoint.cpp
+++ b/libs/full/checkpoint/examples/1d_stencil_4_checkpoint.cpp
@@ -70,13 +70,13 @@ public:
     partition_data() = default;
 
     explicit partition_data(std::size_t size)
-      : data_(std::allocator<double>().allocate(size), size, buffer_type::take)
+      : data_(new double[size], size, buffer_type::take)
       , size_(size)
     {
     }
 
     partition_data(std::size_t size, double initial_value)
-      : data_(std::allocator<double>().allocate(size), size, buffer_type::take)
+      : data_(new double[size], size, buffer_type::take)
       , size_(size)
     {
         double const base_value = static_cast<double>(initial_value * size);
@@ -85,8 +85,7 @@ public:
     }
 
     partition_data(partition_data const& old_part)
-      : data_(std::allocator<double>().allocate(old_part.size()),
-            old_part.size(), buffer_type::take)
+      : data_(new double[old_part.size()], old_part.size(), buffer_type::take)
       , size_(old_part.size())
     {
         for (std::size_t i = 0; i < old_part.size(); i++)

